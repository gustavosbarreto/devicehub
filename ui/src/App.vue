<template>
<v-app>
    <v-navigation-drawer v-if="isLoggedIn" :clipped="false" fixed v-model="drawer" :mini-variant="false" enable-resize-watcher app>
        <v-container>
            <div class="text-center">
                <v-icon>mdi-console</v-icon>
                <h2 style="font-family: monospace">ShellHub</h2>
                <span class="overline">beta</span>
            </div>
        </v-container>
        <v-list>
            <v-list-item v-for="item in items" :key="item.title" :to="item.path" two-line>
                <v-list-item-action>
                    <v-icon v-text="item.icon"></v-icon>
                </v-list-item-action>

                <v-list-item-content>
                    <v-list-item-title v-text="item.title"></v-list-item-title>
                </v-list-item-content>
            </v-list-item>
        </v-list>
    </v-navigation-drawer>
    <v-app-bar app elevate-on-scroll color="primary" class="pl-3 pr-4" v-if="isLoggedIn">
        <v-menu transition="scale-transition" origin="top left">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-server</v-icon>
                    My Device Fleet
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>
        </v-menu>
        <v-spacer></v-spacer>
        <v-chip>
            <v-icon>help</v-icon>
        </v-chip>
        <v-chip>
            <v-icon>notifications</v-icon>
        </v-chip>
        <v-menu transition="scale-transition" origin="top right">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-account</v-icon>
                    {{ $store.getters["auth/currentUser"] }}
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>

            <v-card>
                <v-list-item three-line>
                    <v-list-item-content>
                        <v-list-item-title class="mb-1">Tenant ID</v-list-item-title>
                        <v-list-item-subtitle>
                            <v-chip>
                                <span>
                                    {{ tenant }}
                                </span>
                                <v-icon right v-clipboard="tenant" v-clipboard:success="() => { copySnack = true; }">mdi-content-copy</v-icon>
                            </v-chip>
                        </v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>

                <v-card-actions>
                    <v-btn small text @click="logout()">
                        Logout
                    </v-btn>
                </v-card-actions>
            </v-card>
        </v-menu>
    </v-app-bar>
    <v-content>
        <v-container class="pa-8" fluid>
            <router-view></router-view>
        </v-container>
        <v-snackbar v-model="copySnack" :timeout=3000>Tenant ID copied to clipboard</v-snackbar>
    </v-content>
</v-app>
</template>

<script>
export default {
  name: "App",

  methods: {
    logout() {
      this.$store.dispatch("auth/logout").then(() => {
        this.$router.push("/login");
      });
    }
  },

  computed: {
    tenant() {
      return this.$store.getters["auth/tenant"];
    },

    isLoggedIn() {
      return this.$store.getters["auth/isLoggedIn"];
    }
  },

  data() {
    return {
      drawer: true,
      clipped: false,
      copySnack: false,
      items: [
        {
          icon: "dashboard",
          title: "Dashboard",
          path: "/"
        },
        {
          icon: "devices",
          title: "Devices",
          path: "/devices"
        },
        {
          icon: "history",
          title: "Sessions",
          path: "/sessions"
        }
      ]
    };
  }
};
</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
