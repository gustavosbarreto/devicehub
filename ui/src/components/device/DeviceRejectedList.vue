<template>
  <fragment>
    <v-card-text class="pa-0">
      <v-data-table
        class="elevation-1"
        :headers="headers"
        :items="getListRejectedDevices"
        data-test="dataTable-field"
        :items-per-page="10"
        :footer-props="{'items-per-page-options': [10, 25, 50, 100]}"
        :server-items-length="getNumberRejectedDevices"
        :options.sync="pagination"
      >
        <template slot="no-data">
          There are no more pending devices
        </template>

        <template #[`item.hostname`]="{ item }">
          <router-link :to="{ name: 'detailsDevice', params: { id: item.uid } }">
            {{ item.name }}
          </router-link>
        </template>

        <template #[`item.info.pretty_name`]="{ item }">
          <DeviceIcon
            :icon-name="item.info.id"
            data-test="deviceIcon-component"
          />
          {{ item.info.pretty_name }}
        </template>

        <template #[`item.request_time`]="{ item }">
          {{ item.last_seen | moment("ddd, MMM Do YY, h:mm:ss a") }}
        </template>

        <template #[`item.actions`]="{ item }">
          <v-menu
            :ref="'menu'+getListRejectedDevices.indexOf(item)"
            offset-y
          >
            <template #activator="{ on, attrs }">
              <v-chip
                color="transparent"
                v-on="on"
              >
                <v-icon
                  small
                  class="icons"
                  v-bind="attrs"
                  v-on="on"
                >
                  mdi-dots-horizontal
                </v-icon>
              </v-chip>
            </template>

            <v-card>
              <v-list-item
                @click="showDeviceAcceptButton(getListRejectedDevices.indexOf(item))"
              >
                <DeviceActionButton
                  :uid="item.uid"
                  action="accept"
                  :show.sync="deviceAcceptButtonShow[getListRejectedDevices.indexOf(item)]"
                  data-test="DeviceActionButtonAccept-component"
                  @update="refresh"
                />
              </v-list-item>

              <v-list-item
                @click="showDeviceRemoveButton(getListRejectedDevices.indexOf(item))"
              >
                <DeviceActionButton
                  :uid="item.uid"
                  action="remove"
                  :show.sync="deviceRemoveButtonShow[getListRejectedDevices.indexOf(item)]"
                  data-test="deviceActionButtonRemove-component"
                  @update="refresh"
                />
              </v-list-item>
            </v-card>
          </v-menu>
        </template>
      </v-data-table>
    </v-card-text>
  </fragment>
</template>

<script>

import DeviceIcon from '@/components/device//DeviceIcon';
import DeviceActionButton from '@/components/device/DeviceActionButton';
import formatDeviceSort from '@/components/filter/object';

export default {
  name: 'DeviceListComponent',

  components: {
    DeviceIcon,
    DeviceActionButton,
  },

  data() {
    return {
      pagination: {},
      deviceAcceptButtonShow: [],
      deviceRemoveButtonShow: [],

      headers: [
        {
          text: 'Hostname',
          value: 'hostname',
          align: 'center',
        },
        {
          text: 'Operating System',
          value: 'info.pretty_name',
          align: 'center',
          sortable: false,
        },
        {
          text: 'Request Time',
          value: 'request_time',
          align: 'center',
          sortable: false,
        },
        {
          text: 'Actions',
          value: 'actions',
          align: 'center',
          sortable: false,
        },
      ],
    };
  },

  computed: {
    getListRejectedDevices() {
      return this.$store.getters['devices/list'];
    },

    getNumberRejectedDevices() {
      return this.$store.getters['devices/getNumberDevices'];
    },
  },

  watch: {
    pagination: {
      handler() {
        this.getRejectedDevices();
      },
      deep: true,
    },
  },

  mounted() {
    this.$store.dispatch('devices/resetListDevices');
  },

  methods: {
    async getRejectedDevices() {
      let sortStatusMap = {};

      sortStatusMap = formatDeviceSort(
        this.pagination.sortBy[0],
        this.pagination.sortDesc[0],
      );

      const data = {
        perPage: this.pagination.itemsPerPage,
        page: this.pagination.page,
        filter: this.$store.getters['devices/getFilter'],
        status: 'rejected',
        sortStatusField: sortStatusMap.field,
        sortStatusString: sortStatusMap.statusString,
      };

      try {
        await this.$store.dispatch('devices/fetch', data);

        this.setArrays();
      } catch (error) {
        if (error.response.status === 403) {
          this.$store.dispatch('snackbar/showSnackbarErrorAssociation');
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceListRejected);
        }
      }
    },

    refresh() {
      this.getRejectedDevices();
    },

    showDeviceAcceptButton(index) {
      this.deviceAcceptButtonShow[index] = this.deviceAcceptButtonShow[index] === undefined
        ? true : !this.deviceAcceptButtonShow[index];
      this.$set(this.deviceAcceptButtonShow, index, this.deviceAcceptButtonShow[index]);

      this.closeMenu(index);
    },

    showDeviceRemoveButton(index) {
      this.deviceRemoveButtonShow[index] = this.deviceRemoveButtonShow[index] === undefined
        ? true : !this.deviceRemoveButtonShow[index];
      this.$set(this.deviceRemoveButtonShow, index, this.deviceRemoveButtonShow[index]);

      this.closeMenu(index);
    },

    setArrays() {
      const numberRejectedDevices = this.getListRejectedDevices.length;

      if (numberRejectedDevices > 0) {
        this.deviceAcceptButtonShow = new Array(numberRejectedDevices).fill(false);
        this.deviceRemoveButtonShow = new Array(numberRejectedDevices).fill(false);
      }
    },

    closeMenu(index) {
      this.$refs[`menu${index}`].isActive = false;
    },
  },
};

</script>

<style scoped>

.list-itens {
  font-family: monospace;
}

.icons{
  margin-right: 4px;
}

</style>
