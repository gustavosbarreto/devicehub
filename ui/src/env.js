export default {
  /* eslint-disable no-param-reassign */
  install(Vue) {
    const env = {
      isHosted: window.env.HOSTED === 'true',
    };

    Vue.env = env;
    Vue.prototype.$env = env;
  },
  /* eslint-enable no-param-reassign */
};
