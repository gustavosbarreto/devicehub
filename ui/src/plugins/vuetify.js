import Vue from 'vue';
import Vuetify from 'vuetify/lib';
import '@mdi/font/css/materialdesignicons.css';
import 'font-logos/assets/font-logos.css';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { fas } from '@fortawesome/free-solid-svg-icons';
import Clipboard from 'v-clipboard';
import Fragment from 'vue-fragment';

//import 'vuetify/dist/vuetify.min.css'
//import '../styles/variables.scss'

Vue.component('font-awesome-icon', FontAwesomeIcon); // Register component globally
library.add(fas); // Include needed icons.

Vue.use(Vuetify);
Vue.use(Clipboard);
Vue.use(Fragment.Plugin);

const DEVICE_ICONS = {
  alpine: 'fl-alpine',
  arch: 'fl-archlinux',
  centos: 'fl-centos',
  coreos: 'fl-coreos',
  debian: 'fl-debian',
  devuan: 'fl-devuan',
  elementary: 'fl-elementary',
  fedora: 'fl-fedora',
  freebsd: 'fl-freebsd',
  gentoo: 'fl-gentoo',
  linuxmint: 'fl-linuxmint',
  mageia: 'fl-mageia',
  manjaro: 'fl-manjaro',
  mandriva: 'fl-mandriva',
  nixos: 'fl-nixos',
  opensuse: 'fl-opensuse',
  rhel: 'fl-redhat',
  sabayon: 'fl-sabayon',
  slackware: 'fl-slackware',
  ubuntu: 'fl-ubuntu',
  raspbian: 'fl-raspberry-pi',
  'ubuntu-core': 'fl-ubuntu',
  void: 'fl-void',
};

export default new Vuetify({
  iconfont: 'md',
  icons:{
    values:DEVICE_ICONS
  },
  theme: {
    dark: false,
    themes: {
      light: {
        primary: '#364576',
        secondary: '#b0bec5',
        accent: '#8c9eff',
        error: '#b71c1c',
      },
    }
  }
});
