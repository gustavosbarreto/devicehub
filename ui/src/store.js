import Vue from 'vue';
import Vuex from 'vuex';
import stats from '@/modules/stats';
import sessions from '@/modules/sessions';
import auth from '@/modules/auth';
import devices from '@/modules/devices';
import modals from '@/modules/modals';
import firewallrules from '@/modules/firewall_rules';
import notifications from '@/modules/notifications';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    devices, modals, stats, sessions, auth, firewallrules, notifications,
  },
});
