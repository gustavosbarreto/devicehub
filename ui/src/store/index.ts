import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";

import { auth, AuthState } from "./modules/auth";
import { apiKeys, ApiKeysState } from "./modules/api_keys";
import { layout, LayoutState } from "./modules/layout";
import { users, UsersState } from "./modules/users";
import { tags, TagsState } from "./modules/tags";
import { stats, StatsState } from "./modules/stats";
import { spinner, SpinnerState } from "./modules/spinner";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { sessions, SessionsState } from "./modules/sessions";
import { security, SecurityState } from "./modules/security";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { privateKey, PrivateKeyState } from "./modules/private_key";
import { notifications, NotificationsState } from "./modules/notifications";
import { modal, ModalState } from "./modules/modal";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { devices, DevicesState } from "./modules/devices";
import { box, BoxState } from "./modules/box";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { billing } from "./modules/billing";
import { customer, CustomerState } from "./modules/customer";
import { announcement, AnnouncementState } from "./modules/announcement";
import apiPlugin from "./plugins/api";

export interface State {
  auth: AuthState;
  apiKeys: ApiKeysState;
  billing: NamespacesState;
  customer: CustomerState;
  box: BoxState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  modal: ModalState;
  namespaces: NamespacesState;
  notifications: NotificationsState;
  privateKey: PrivateKeyState;
  publicKeys: PublicKeysState;
  security: SecurityState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  tags: TagsState;
  users: UsersState;
  announcement: AnnouncementState;
}

export const key: InjectionKey<Store<State>> = Symbol("store");

export const store = createStore<State>({
  modules: {
    auth,
    apiKeys,
    customer,
    billing,
    box,
    devices,
    firewallRules,
    layout,
    modal,
    namespaces,
    notifications,
    privateKey,
    publicKeys,
    security,
    sessions,
    snackbar,
    spinner,
    stats,
    tags,
    users,
    announcement,
  },
  plugins: [
    apiPlugin,
  ],
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
