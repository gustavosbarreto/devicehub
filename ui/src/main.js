import Vue from 'vue';

import Snackbar from '@/components/snackbar/Snackbar';

import App from './App';
import router from './router';
import store from './store';
import env from './env';
import './vee-validate';
import vuetify from './plugins/vuetify';

Vue.config.productionTip = false;

Vue.component('Snackbar', Snackbar);

Vue.use(require('vue-moment'));

Vue.use(env);

new Vue({
  vuetify,
  router,
  store,
  render: (h) => h(App),
}).$mount('#app');
