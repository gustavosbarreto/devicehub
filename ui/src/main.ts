import { createApp } from "vue";
import * as Sentry from "@sentry/vue";
import { BrowserTracing } from "@sentry/tracing";
import { envVariables } from "./envVariables";
import vuetify from "./plugins/vuetify";
import { key, store } from "./store";
import { router } from "./router";
import App from "./App.vue";

import { loadFonts } from "./plugins/webfontloader";

import SnackbarComponent from "./components/Snackbar/Snackbar.vue";
import { SnackbarPlugin } from "./plugins/snackbar";

const app = createApp(App);

Sentry.init({
  app,
  dsn: envVariables.sentryDsn || "",
  integrations: [
    new BrowserTracing({
      routingInstrumentation: Sentry.vueRouterInstrumentation(router),
    }),
  ],
  tracesSampleRate: 1.0,
  hooks: ["activate", "create", "destroy", "mount", "update"],
  timeout: 500,
  release: envVariables.version || "latest",
});
Sentry.setTag("project", "shellhub-ui");

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.use(SnackbarPlugin);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
