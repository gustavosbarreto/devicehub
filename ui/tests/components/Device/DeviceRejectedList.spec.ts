/* eslint-disable @typescript-eslint/no-empty-function */
import { createVuetify } from "vuetify";
import { mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it } from "vitest";
import { createStore } from "vuex";
import DeviceRejectedList from "../../../src/components/Devices/DeviceRejectedList.vue";
import { key } from "../../../src/store";
import routes from "../../../src/router";

const numberDevices = 2;
const devicesSelected = [];
const action = "suggestedDevices";

const headers = [
  {
    text: "Hostname",
    value: "name",
    sortable: true,
  },
  {
    text: "Operating System",
    value: "operating_system",
  },
  {
    text: "Request Time",
    value: "request_time",
  },
  {
    text: "Actions",
    value: "actions",
  },
];

const devices = [
  {
    uid: "a582b47a42d",
    name: "39-5e-2a",
    identity: {
      mac: "00:00:00:00:00:00",
    },
    info: {
      id: "linuxmint",
      pretty_name: "Linux Mint 19.3",
      version: "",
    },
    public_key: "----- PUBLIC KEY -----",
    tenant_id: "00000000",
    last_seen: "2020-05-20T18:58:53.276Z",
    online: false,
    namespace: "user",
    status: "accepted",
  },
  {
    uid: "a582b47a42e",
    name: "39-5e-2b",
    identity: {
      mac: "00:00:00:00:00:00",
    },
    info: {
      id: "linuxmint",
      pretty_name: "Linux Mint 19.3",
      version: "",
    },
    public_key: "----- PUBLIC KEY -----",
    tenant_id: "00000001",
    last_seen: "2020-05-20T19:58:53.276Z",
    online: true,
    namespace: "user",
    status: "accepted",
  },
];

const store = createStore({
  state: {
    devices,
    numberDevices,
    devicesSelected,
  },
  getters: {
    "devices/getDevicesForUserToChoose": (state) => state.devices,
    "devices/getNumberForUserToChoose": (state) => state.numberDevices,
    "devices/getDevicesSelected": (state) => state.devicesSelected,
    "devices/getNumberDevices": (state) => state.numberDevices,
    "devices/list": (state) => state.devices,
  },
  actions: {
    "devices/fetch": () => {},
    "devices/setDevicesSelected": () => {},
    "devices/getDevicesMostUsed": () => {},
    "devices/setDevicesForUserToChoose": () => {},
    "devices/getNumberForUserToChoose": () => {},
    "devices/resetListDevices": () => {},
    "snackbar/showSnackbarCopy": () => {},
    "snackbar/showSnackbarErrorAssociation": () => {},
    "snackbar/showSnackbarErrorLoading": () => {},
  },
});

describe("Devices Pending List", () => {
  let wrapper: VueWrapper<InstanceType<typeof DeviceRejectedList>>;

  beforeEach(() => {
    const vuetify = createVuetify();
    wrapper = mount(DeviceRejectedList, {
      global: {
        plugins: [[store, key], vuetify, routes],
        stubs: ["router-link", "router-view"],
      },
      props: {
        action,
      },
      data() {
        return {
          headers,
          devices,
        };
      },
      shallow: true,
    });
  });

  ///////
  // Component Rendering
  //////

  it("Is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });
  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Compare data with default value", () => {
    expect(wrapper.vm.headers).toEqual(headers);
    expect(wrapper.vm.devices).toEqual(devices);
    expect(wrapper.vm.numberDevices).toEqual(numberDevices);
  });

  it("Process data in methods", () => {
    const address = `${devices[0].namespace}.${devices[0].name}@localhost`;
    expect(wrapper.vm.sshidAddress(devices[0])).toEqual(address);
  });

  //////
  // HTML validation
  //////

  it("Renders the template with components", async () => {
    expect(wrapper.find('[data-test="devices-list"]').exists()).toEqual(
      true,
    );
  });
  it("Renders the template with data", () => {
    const dataTable = wrapper.find('[data-test="devices-list"]');
    const dataTableAttr = dataTable.attributes();
    expect(dataTable.exists()).toEqual(true);
    expect(+dataTableAttr.totalcount).toBe(numberDevices);
    expect(+dataTableAttr.actualpage).toBe(1);
  });
});
