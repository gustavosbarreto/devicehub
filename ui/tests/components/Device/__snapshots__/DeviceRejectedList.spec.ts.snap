// Vitest Snapshot v1

exports[`Devices Pending List > Renders the component 1`] = `
"<div action=\\"suggestedDevices\\" data-v-dede1877=\\"\\">
  <data-table-stub itemsperpage=\\"10\\" totalcount=\\"2\\" headers=\\"[object Object],[object Object],[object Object],[object Object]\\" items=\\"[object Object],[object Object]\\" comboboxoptions=\\"10,20,50,100\\" loading=\\"false\\" actualpage=\\"1\\" nextpage=\\"async () => {
      await getDevices(itemsPerPage.value, ++page.value);
    }\\" previouspage=\\"async () => {
      try {
        if (page.value > 1)
          await getDevices(itemsPerPage.value, --page.value);
      } catch (error) {
        store.dispatch(&quot;snackbar/setSnackbarErrorDefault&quot;);
        throw new Error(error);
      }
    }\\" data-test=\\"devices-list\\" data-v-dede1877=\\"\\"></data-table-stub>
</div>"
`;
